pkg/tests/upgrade_tests/upgrade_test.go

```diff
@@ -635,3 +635,213 @@ func copyMetrics(metrics []prompb.TimeSeries) []prompb.TimeSeries {
+
+		var version string
+		db, err = pgx.Connect(ctx, testhelpers.PgConnectURL("postgres", testhelpers.Superuser))
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = db.QueryRow(ctx, `SELECT extversion FROM pg_extension where extname='timescaledb'`).Scan(&version)
+		if err != nil {
+			t.Fatal(err.Error())
+		}
+		db.Close(ctx)
+		if version != "2.0.0-rc4" {
+			t.Fatal("failed to verify upgrade extension with -upgrade-prerelease-extension true")
+		}
+		t.Logf("successfully tested extension upgrade flow with --upgrade-prereleases-extensions true")
+	}()
+
+	os.Exit(0)
+}
+
+func TestExtensionUpgrade(t *testing.T) {
+	var err error
+	var version string
+	ctx := context.Background()
+
+	db, dbContainer, closer := startDB(t, ctx)
+	defer testhelpers.StopContainer(ctx, dbContainer, false)
+
+	defer closer.Close()
+
+	// start promscale & test upgrade-extensions as false
+	func() {
+		connectorImage := "vineeth97/promscale:t1"
+		databaseName := "postgres"
+		connector, err := testhelpers.StartConnectorWithImage(ctx, connectorImage, *printLogs, []string{}, databaseName)
+		if err != nil {
+			t.Fatal(err.Error())
+		}
+		defer testhelpers.StopContainer(ctx, connector, *printLogs)
+
+		err = db.QueryRow(ctx, `SELECT extversion FROM pg_extension where extname='timescaledb'`).Scan(&version)
+		if err != nil {
+			t.Fatal(err.Error())
+		}
+
+		if version != "2.0.0-rc4" {
+			t.Fatal("failed to verify upgrade extension with -upgrade-prerelease-extension false")
+		}
+	}()
+
+	extVersion := "1.7.3"
+	dropAndCreateExt(t, ctx, extVersion)
+
+	db, err = pgx.Connect(ctx, testhelpers.PgConnectURL("postgres", testhelpers.Superuser))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = db.QueryRow(ctx, `SELECT extversion FROM pg_extension where extname='timescaledb'`).Scan(&version)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	db.Close(ctx)
+	if version != extVersion {
+		t.Fatal("failed to verify upgrade extension with -upgrade-prerelease-extension false")
+	}
+
+	// start a new connector and test --upgrade-extensions as true which is by default set in flags.
+	func() {
+
+		connectorImage := "vineeth97/promscale:t1"
+		databaseName := "postgres"
+		connector, err := testhelpers.StartConnectorWithImage(ctx, connectorImage, *printLogs, []string{}, databaseName)
+		if err != nil {
+			t.Fatal(err.Error())
+		}
+		defer testhelpers.StopContainer(ctx, connector, *printLogs)
+
+		var version string
+		db, err = pgx.Connect(ctx, testhelpers.PgConnectURL("postgres", testhelpers.Superuser))
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = db.QueryRow(ctx, `SELECT extversion FROM pg_extension where extname='timescaledb'`).Scan(&version)
+		if err != nil {
+			t.Fatal(err.Error())
+		}
+		db.Close(ctx)
+		// Note this needs to be fixed the expected version is 1.7.4 but
+		// db is failing to update with an error ERROR: extension "timescaledb" cannot be updated after the old version has already been loaded
+		if version != "1.7.3" {
```

Comment:
@cevian I noticed an issue when writing the e2e test. By default, if timescaleDB installed version is 1.7.3 & the timescaleDB image has now 1.7.4 & 2.0.0 version available.

The expected behavior is promscale on startup should upgrade timescaleDB version from `1.7.3` to `1.7.4` . But while performing this DB returns an error  `ERROR: extension "timescaledb" cannot be updated after the old version has already been loaded` not sure about the exact reason as this is from tsdb. I am also verified promscale is creating a new conn before peforming ext upgrade. 

This isn't a blocker for PR to merge. But the e2e case can be fulfilled with the workaround.

Reply:
This seems like a big issue I'll have to look into it

Reply:
hmm when I run this into this it seems like the error is actually "ERROR:  extension "timescaledb" has no update path from version "1.7.3" to version "1.7.4"  on the db side. And that's because the only upgrade path in the docker image `vineeth97/timescaledb-ext-test` is TO 2.0.0-rc4 so 1.7.3-> 2.0.0-rc4 not 1.7.3->1.7.4 (we usually only include upgrade scripts to one version in our docker images). Perhaps we should use another image for this test?